From a218daddac8e1e4ef8cfaa8c535ccc95f8ab7427 Mon Sep 17 00:00:00 2001
From: David Yu Yang <davidyang6us@gmail.com>
Date: Mon, 25 Oct 2021 22:29:24 +0800
Subject: [PATCH] Fix system font only being applied to last tab

Resolves #1947
---
diff --git a/guake/gsettings.py b/guake/gsettings.py
index 965a2273..9ece43be 100644
--- a/guake/gsettings.py
+++ b/guake/gsettings.py
@@ -377,7 +377,13 @@ def fstyle_changed(self, settings, key, user_data):
         else:
             terminals = self.guake.notebook_manager.iter_terminals()
 
-        font = Pango.FontDescription(settings.get_string(key))
+        if self.settings.general.get_boolean("use-default-font"):
+            gio_settings = Gio.Settings("org.gnome.desktop.interface")
+            font_name = gio_settings.get_string("monospace-font-name")
+        else:
+            font_name = settings.get_string(key)
+
+        font = Pango.FontDescription(font_name)
         for i in terminals:
             i.set_font(font)
 
