--- a/guake/utils.py
+++ b/guake/utils.py
@@ -165,11 +165,11 @@ class FullscreenManager():
         self.window_state = window_state
         setattr(self.window, self.FULLSCREEN_ATTR, bool(window_state & Gdk.WindowState.FULLSCREEN))
 
-        if not window_state & Gdk.WindowState.WITHDRAWN:
-            if self.is_fullscreen():
-                self.fullscreen()
-            else:
-                self.unfullscreen()
+        #if not window_state & Gdk.WindowState.WITHDRAWN:
+        #    if self.is_fullscreen():
+        #        self.fullscreen()
+        #    else:
+        #        self.unfullscreen()
 
     def fullscreen(self):
         self.window.fullscreen()
