diff -rupN cinnamon-screensaver-2.6.4.orig/configure.ac cinnamon-screensaver-2.6.4/configure.ac
--- cinnamon-screensaver-2.6.4.orig/configure.ac	2015-06-26 08:15:54.000000000 -0400
+++ cinnamon-screensaver-2.6.4/configure.ac	2015-09-25 15:02:31.148892579 -0400
@@ -58,7 +58,7 @@ PKG_CHECK_MODULES(CINNAMON_SCREENSAVER,
         gtk+-3.0 >= $GTK_REQUIRED_VERSION
         dbus-glib-1 >= $DBUS_REQUIRED_VERSION
         cinnamon-desktop >= $CINNAMON_DESKTOP_REQUIRED_VERSION)
-PKG_CHECK_MODULES([WEBKIT], [webkit2gtk-3.0],, [PKG_CHECK_MODULES([WEBKIT], [webkit2gtk-4.0])])
+
 AC_SUBST(CINNAMON_SCREENSAVER_CFLAGS)
 AC_SUBST(CINNAMON_SCREENSAVER_LIBS)
 
@@ -422,6 +422,22 @@ if test "$have_xf86gamma" = yes -o "$hav
 fi
 
 dnl ---------------------------------------------------------------------------
+dnl - Optional webkit support
+dnl ---------------------------------------------------------------------------
+
+AC_ARG_ENABLE(webkit,[
+GTK Webkit support:
+
+  --enable-webkit         Enable support for GTK webkit screensavers.
+  --disable-webkit        Disable support for GTK webkit screensavers.],
+[enable_webkit="$enableval"],[enable_webkit=yes])
+
+if test "x$enable_webkit" = xyes; then
+    PKG_CHECK_MODULES([WEBKIT], [webkit2gtk-3.0],, [PKG_CHECK_MODULES([WEBKIT], [webkit2gtk-4.0])])
+fi
+AM_CONDITIONAL([HAS_WEBKIT], [test "x$enable_webkit" = xyes])
+
+dnl ---------------------------------------------------------------------------
 dnl - The --enable-locking option
 dnl ---------------------------------------------------------------------------
 
@@ -752,6 +768,7 @@ echo "
         Maintainer mode:          ${USE_MAINTAINER_MODE}
         Docs enabled:             ${enable_docbook_docs}
 
+        GTK WebKit screensavers:  ${enable_webkit}
         Screen locking enabled:   ${enable_locking}
         Show keyboard indicator:  ${with_kbd_layout_indicator}
 
diff -rupN cinnamon-screensaver-2.6.4.orig/data/screensavers/Makefile.am cinnamon-screensaver-2.6.4/data/screensavers/Makefile.am
--- cinnamon-screensaver-2.6.4.orig/data/screensavers/Makefile.am	1969-12-31 19:00:00.000000000 -0500
+++ cinnamon-screensaver-2.6.4/data/screensavers/Makefile.am	2015-09-25 21:44:56.525096815 -0400
@@ -0,0 +1,16 @@
+screensaversdir = $(pkgdatadir)/screensavers
+
+xscreensaver = xscreensaver@cinnamon.org
+xscreensaverdir = $(screensaversdir)/$(xscreensaver)
+dist_xscreensaver_SCRIPTS = $(xscreensaver)/main
+
+if HAS_WEBKIT
+webkit = webkit@cinnamon.org
+webkitdir = $(screensaversdir)/$(webkit)
+dist_webkit_SCRIPTS = $(webkit)/main
+
+webkit_stars = $(webkit)/webkit-stars@cinnamon.org
+webkit_starsdir = $(screensaversdir)/$(webkit_stars)
+dist_webkit_stars_DATA = $(webkit_stars)/index.html $(webkit_stars)/metadata.json
+endif
+
diff -rupN cinnamon-screensaver-2.6.4.orig/data/screensavers/Makefile.in cinnamon-screensaver-2.6.4/data/screensavers/Makefile.in
--- cinnamon-screensaver-2.6.4.orig/data/screensavers/Makefile.in	2015-06-26 08:15:54.000000000 -0400
+++ cinnamon-screensaver-2.6.4/data/screensavers/Makefile.in	1969-12-31 19:00:00.000000000 -0500
@@ -1,15 +0,0 @@
-prefix = @prefix@
-datadir = @datadir@
-pkgdatadir = $(datadir)/@PACKAGE@
-all:
-
-clean:
-
-distclean:
-
-install:
-	mkdir -p $(DESTDIR)$(pkgdatadir)/screensavers/; \
-	find -mindepth 1 -maxdepth 1 -type d -exec cp -R {} $(DESTDIR)$(pkgdatadir)/screensavers/ \;
-
-uninstall:
-	find -mindepth 1 -type f -exec rm $(DESTDIR)$(pkgdatadir)/screensavers/{} \;
